File: benchmarks/fixtures/webapp_rb/database/migrations.rb
# Database migration management.

require_relative '../utils/helpers'
require_relative 'connection'
require_relative '../exceptions'

MIGRATIONS = [
  { version: '001', name: 'create_users', sql: 'CREATE TABLE users (id TEXT PRIMARY KEY, email TEXT, name TEXT, role TEXT)' },
  { version: '002', name: 'create_sessions', sql: 'CREATE TABLE sessions (id TEXT PRIMARY KEY, user_id TEXT, token_hash TEXT)' },
  { version: '003', name: 'create_payments', sql: 'CREATE TABLE payments (id TEXT PRIMARY KEY, user_id TEXT, amount REAL)' },
  { version: '004', name: 'create_events', sql: 'CREATE TABLE events (id TEXT PRIMARY KEY, type TEXT, payload TEXT)' },
  { version: '005', name: 'create_notifications', sql: 'CREATE TABLE notifications (id TEXT PRIMARY KEY, user_id TEXT, channel TEXT)' },
].freeze

# Run pending migrations.
class MigrationRunner
  def initialize(db)
    @db = db
    get_logger('database.migrations').info('MigrationRunner initialized')
  end

  def run_pending
    count = 0
    MIGRATIONS.each do |migration|
      get_logger('database.migrations').info("Applying migration #{migration[:version]}: #{migration[:name]}")
      begin
        @db.begin_transaction
        @db.execute_query(migration[:sql])
        @db.commit
        count += 1
      rescue StandardError => e
        @db.rollback
        raise DatabaseError.new("Migration #{migration[:version]} failed: #{e}")
      end
    end
    get_logger('database.migrations').info("#{count} migrations applied")
    count
  end

  def status
    { applied: 0, pending: MIGRATIONS.length, total: MIGRATIONS.length }
  end
end

File: benchmarks/fixtures/webapp_ts/src/database/migrations.ts
/** Database migration management. */
import { getLogger } from '../utils/helpers';
import { DatabaseConnection } from './connection';
import { DatabaseError } from '../errors';

const logger = getLogger("database.migrations");

interface Migration {
    version: string;
    name: string;
    sql: string;
}

const MIGRATIONS: Migration[] = [
    { version: "001", name: "create_users", sql: "CREATE TABLE users (id TEXT PRIMARY KEY, email TEXT, name TEXT, role TEXT)" },
    { version: "002", name: "create_sessions", sql: "CREATE TABLE sessions (id TEXT PRIMARY KEY, userId TEXT, tokenHash TEXT)" },
    { version: "003", name: "create_payments", sql: "CREATE TABLE payments (id TEXT PRIMARY KEY, userId TEXT, amount REAL)" },
    { version: "004", name: "create_events", sql: "CREATE TABLE events (id TEXT PRIMARY KEY, type TEXT, payload TEXT)" },
    { version: "005", name: "create_notifications", sql: "CREATE TABLE notifications (id TEXT PRIMARY KEY, userId TEXT, channel TEXT)" },
];

/** Run pending migrations. */
export class MigrationRunner {
    constructor(private db: DatabaseConnection) {
        logger.info("MigrationRunner initialized");
    }

    async runPending(): Promise<number> {
        let count = 0;
        for (const migration of MIGRATIONS) {
            logger.info(`Applying migration ${migration.version}: ${migration.name}`);
            try {
                this.db.beginTransaction();
                await this.db.executeQuery(migration.sql);
                this.db.commit();
                count++;
            } catch (e) {
                this.db.rollback();
                throw new DatabaseError(`Migration ${migration.version} failed: ${e}`);
            }
        }
        logger.info(`${count} migrations applied`);
        return count;
    }

    async status(): Promise<{ applied: number; pending: number; total: number }> {
        return { applied: 0, pending: MIGRATIONS.length, total: MIGRATIONS.length };
    }
}
